addons/mail/models/mail_alias_domain.py

```diff
@@ -0,0 +1,99 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, exceptions, fields, models, _
+
+
+class AliasDomain(models.Model):
+    """ Model alias domains, now company-specific. Alias domains are email
+    domains used to receive emails through catchall and bounce aliases, as
+    well as using mail.alias records to redirect email replies.
+
+    This replaces ``mail.alias.domain`` configuration parameter use until v16.
+    """
+    _name = 'mail.alias.domain'
+    _description = "Email Domain"
+    _order = 'sequence ASC, id DESC'
+    _rec_name = 'name'
+
+    name = fields.Char('Name', required=True)
+    company_ids = fields.One2many(
+        'res.company', 'alias_domain_id', string='Used In')
+    sequence = fields.Integer(default=10)
+    bounce = fields.Char('Bounce Alias', default='bounce', required=True)
+    bounce_email = fields.Char('Bounce Email', compute='_compute_bounce_email')
+    catchall = fields.Char('Catchall Alias', default='catchall', required=True)
+    catchall_email = fields.Char('Catchall Email', compute='_compute_catchall_email')
+
+    @api.depends('bounce', 'name')
+    def _compute_bounce_email(self):
+        self.bounce_email = ''
+        for domain in self.filtered('bounce'):
+            domain.bounce_email = f'{domain.bounce}@{domain.name}'
+
+    @api.depends('catchall', 'name')
+    def _compute_catchall_email(self):
+        self.catchall_email = ''
+        for domain in self.filtered('catchall'):
+            domain.catchall_email = f'{domain.catchall}@{domain.name}'
+
+    _sql_constraints = [
```

Comment:
as discussed on discord, it can be cool if the constraint was on `bounce_email` and `catchall_email` instead :) (we can store the fields, or just making a @api.depends, both are ok for me)